### ISSUE TITLE
Request for guides to be added to Fluent documentation on how to integrate with notable libraries

### ISSUE THREAD

**Issue by mattmazzola** [microsoft/fluentui#21481]
This request for documentation change not a technical feature change. It didn't seem to fit under any of the other issue categories/templates so I chose Feature Request as it seemed closest.

#### Describe the feature that you would like added

Not a technical feature, but perhaps only documentation change if it works without requiring technical changes.

Similar to how NextJS setup is described here:
https://github.com/microsoft/fluentui/wiki/Server-side-rendering-and-browserless-testing#nextjs-setup

I thought there could be a new section added for [Remix](https://remix.run/docs/en/v1) since the techniques are likely solving similar problems but in different ways based on the library/framework.

We are staring a new internal project and were attempting to use [Remix](https://remix.run/docs/en/v1) instead of more standard React app and we're looking to use FluentUI to maintain consistent feeling with our other app design. We ran into similar issues as seen in this open item and it caused me to open this issue.
https://github.com/microsoft/fluentui/issues/11411

Note the last comment in that linked issue also asking about Remix so it seems there is interest.

The existing [Getting Started](https://developer.microsoft.com/en-us/fluentui#/get-started/web) does not include information on integrating with these libraries such as Next or Remix, but given their increase in option and this is a much more common starting point I thought there could be a link to these detailed Next or Remix guides.

#### What component or utility would this be added to

N/A

#### Have you discussed this feature with our team, and if so, who

N/A

#### Additional context/screenshots

N/A

**Comment by Promhize** [comment]
@mattmazzola Did you find a way to integrate with Remix-run? It's currently a blocking issue for me

**Comment by paulgildea** [comment]
@mattmazzola Thanks for the feedback. I'm not terribly familiar with Remix, but would be open to a contribution to our wiki on how to use it with Fluent UI React if you have experience with it.

**Comment by mattmazzola** [comment]
> I'm not terribly familiar with Remix [...] you have experience with it

It's new so it's understandable that many don't have experience. I was lucky we were starting a new project and were able to use that opportunity to try it. I think I have enough experience with Remix to help with the guide but would need knowledge from someone who understands FluentUI to know which parts would need work arounds to know what to put into the guide.

I think it is part of Remix philosophy is to have an app which works without Javascript but may become better by adding Javascript.  

For example, two main problems that I ran into were related to Icons and Modals

Fluents uses `initializeIcons` to load icons using Javascript. This would have to be worked around likely be directly linking to the CDN to make the fonts available.
https://github.com/microsoft/fluentui/wiki/Using-icons#font-based-icons

Also, the modals do add the DOM to the modal zone element but the backdrop does not show.

I think we would want someone who is familiar with the problems overcome in the Next guide for Fluent to understand Remix to know if the issue would be the same for not. The implementation of the fix will be different since Remix has different constructs should be the same.  I'm not familiar with Next so I wasn't able to do this myself. 

> [I] would be open to a contribution to our wiki

What is this process of updating the wiki like? Do we simply submit a pull request?
I'm not sure when I would get time to look into this again, but hopefully within next month

**Comment by shuebner20** [comment]
TL;DR; While exciting both FluentUI and Remix, I've learned that it's a painful job to combine them. Maybe it's impossible without negating some core principles of FluentUI or Remix.

# My story

I'm also trying hard to implement FluentUI React in a Remix app. I've done several steps based on the documented steps for Next.JS to get it done. Some of the issues are solved, some are not. After several days of trying and tweaking, I'm afraid that FluentUI's logic does not fully match the Remix principles.

**Disclaimer:** As Remix is a new framework with an [interesting new approach](https://remix.run/docs/en/v1/pages/philosophy) (IMHO), I'm also new to this framework. Please take some Mea Culpa if I'd understood or recite something wrong.

Other than Next.JS (for example), Remix is heavily based on the Fetch API, SSR and [hydration](https://reactjs.org/docs/react-dom.html#hydrate) on the client. Basically every single part of a Remix-based web site or SPA will render on the server by default and is delivered in chunks to the client. Another important difference is that Remix will render (and therefore hydrate) the **full** site. There is no global index.[html|js|ts], app.[js|ts] or similar file which represents the HTML document skeleton including something like a `<div id="root"><!-- Your app goes here --></div>` as it is common practice nowadays with `ReactDOM.render`.

Main issue is that the FluentUI React package is doing a lot of internal stuff upon rendering. This includes:

1. `ComboBox` not rendered as `select` with `options` by default
2. FluentUI expects an empty `<style data-merge-styles="true"></style>` in the document's `head` and causes a hydration warning of fully SSR built `html` document
3. CSS rules will be collected and autogenerated on rendering, limited to the rules required by the rendered controls
4. FluentUI uses internal counters to name (autogenerated) class names and element names, causing mismatches between SSR and client DOMs.

# More details

I'll try to give some additional information based on my experiences.

## `ComboBox` not rendered as `select` with `options` by default

According to the [documentation](https://developer.microsoft.com/en-us/fluentui#/controls/web/combobox#accessibility), the FluentUI''s `ComboBox` component will use a custom layer to render the popup and its options instead of adding `option` elements. Differing to the Remix principle, that the output will render and work correctly in client browser even if JavaScript is disabled or your network connection is poor (Hello Germany *smile*), FluentUI seems to rely on both to render an interactive combo box.

A possible solution could be to disable the custom popup layer for every instance of `ComboBox` as described in the documentation linked above. However, I've not tried it yet and am not sure about the UX consequences.

Please note, that other basic inputs (`TextField`, for example) will work as plain `input` elements in Remix without JavaScript or on throttled connections. There is no theming / styling at all, but you're able to see and edit the value and submit the changes to the Remix server.

This issue might be related to the issue with modals, as @mattmazzola mentioned.

## FluentUI expects an empty `<style data-merge-styles="true"></style>` in the document's `head`

FluentUI expects an empty `<style data-merge-styles="true"></style>` in the document's `head` ([reference](https://github.com/microsoft/fluentui/blob/c3347c67d24f8c51fa9a10377ee64413fac34fb4/packages/merge-styles/README.md#controlling-where-styles-are-injected)). If it is not present, it will be automatically inserted to the document's `head`. This will also occur on SSR and therefore [causes a warning](https://reactjs.org/docs/react-dom.html#hydrate) in `ReactDOM.hydrate` (in dev mode) which might cause other (remember, Remix renders the full `html` document on server side, and hydrates it on the client afterwards). See also this [SO question](https://stackoverflow.com/questions/70457672/how-to-get-ssr-to-work-with-fluentui-remix-run). 

If you want to stay with hydration with  (one of the primary benefits of Remix) you have to make sure, that the autogenerated `style` element is removed before the SSR response is hydrated at the client. One possibility is to remove the element immediately before `ReactDOM.hydrate` is called. This might be done in the client's entry point (`entry.client.tsx`):

```
import { hydrate } from "react-dom";
import { RemixBrowser } from "remix";

export function hydrateClient(): void {
  if (typeof document !== "undefined") {
    if (typeof document.head !== undefined) {
      for (let i = 0; i < document.head.childNodes.length; i++) {
        const childNode = document.head.childNodes[i];
        if (childNode.nodeName === "STYLE") {
          
          console.log("Not good, there is a STYLE element in the HEAD element.");
          // remove the child or hydrate will fail
          childNode.remove();
        }
      }
    }
    // console.log(document.head.innerHTML);
  }

  hydrate(<RemixBrowser />, document);
}

// in entry.client.tsx replace...
// hydrate(<RemixBrowser />, document);
// ... with the new method
hydrateClient();
```

You might also notice in the answer of the linked SO question, that they no longer use `ReactDOM.hydrate`.

## CSS rules will be collected and autogenerated on rendering, limited to the rules required by the rendered controls

In FluentUI, CSS rules will be collected and autogenerated on rendering (my assumption), limited to the rules required by the rendered controls. As Remix will only render exactly that part of the application that need to be refreshed, the CSS rules extracted by [renderStatic](https://github.com/microsoft/fluentui/tree/master/packages/merge-styles#server-side-rendering) will have a limited extent.

As an example for this Remix principle think of inline edit control. In your app you have a summary page or dashboard, displaying a lot of configuration values using `Label`s. Clicking the Pen icon next to any displayed value will exchange this value's `Label` with a `TextField` to allow quick edit. When submitting your change, Remix will post the data as FormData using Fetch API. The server part will process the input (e.g. storing to database) and Remix' SSR will render just the `Label` of the changed value, as this is the only part within the DOM which needs to be updated to reflect the current data. In this example, `renderStatic` will not return any rules but these which are generated during rendering a single `Label` control.

Furthermore, the CSS rules will not be available before SSRing the content (as they are collected during render process and can depend on component's props). While it seems easy to fetch the generated CSS rules for every request in `entry.server.tsx` and insert/append it into the generated DOM string, this excludes most of Remix built-in logic to capture this generated styles neither as integral part of the rendered component nor of the Remix application itself. If I understand correctly, you'll loose some of the fancy response-payload-hash-caching (a.k.a. the performance™). And with adding `style` elements to the SSR DOM string (and thus modifying it after if leaves Remix's SSR pipeline), you're bypassing contents to the client's DOM which the client part Remix might not be aware of, possibly followed by hydration issues. 

Having this, it's pretty difficult to get all of the generated styles from the server to the client after every request. Especially, I'm wondering if the DOM will be blown up by subsequently inserted `style` elements. I guess client's FluentUI logic might not be aware of these styles, as they are generated (moreover named) at the server, so it will not be possible to reuse or free styles as intended.

Thinking of a scaling Remix application running on the edge (not the browser), behind load balancers or dynamic proxies serving several and changing clients on one hand and autogenerated CSS rules, class and element names (which might also be reused within a sever process) on the other hand, this issue could be more complex as one might think first.

This issue is also affected by the following one. So let's seamlessly continue with... 

## FluentUI uses internal counters to name (autogenerated) class names and element names

FluentUI uses internal counters to name (autogenerated) class names and element names or ids, causing mismatches between SSR and client DOMs. It's almost impossible to keep the counters in sync between the server and the client (even with [resetIds](https://docs.microsoft.com/en-us/javascript/api/utilities?view=office-ui-fabric-react-latest#@fluentui-utilities-resetids-function(1))) as Remix will create small UI bundles instead of the full application for each request. This will result in mismatches in class and element names between the SSR and the hydrated client DOM.

If this issue remain unresolved, most likely CSS rules will not be applied as expected and some UX features might not work (such as the modal overlay or the combo box popup).

As mentioned above, this issue is able to affect principles such as reuse and free styles.

# The Solution?

After some days of hard work, facing a lot of problems and succeeding some of them, there are some unresolved issues as of now. Thinking of business grade applications (that's why we choose FluentUI, isn't it?) there are several aspects to be understood before even setting up your CI build pipeline (like the CSS rules and the IDs). I bet that it will not be possible to bundle a successful FluentUI Remix application. At least, I would recommend not to try this at home ;-)

# The Result

For now, I respect the differing principles of FluentUI and Remix, and opted-out of SSR :-/

Nonetheless, I have this dream... Both FluentUI and Remix are (or will be) core frameworks for our SPA B2B applications. I'll definitively watch this topic and give it another try on the very-next opportunity.

Meanwhile, if you want to follow my example, just wrap your FluentUI application with the `ClientHost` component from the remix-utils package (need to be installed additionally to Remix) to disable SSR. As simplest, wrap the `Outlet` component in the default `root.tsx`:

```
import { ClientOnly } from "remix-utils";

<!-- snip -->
export default function App() {
  return (
    <html lang="en">
      <head>
        <meta charSet="utf-8" />
        <meta name="viewport" content="width=device-width,initial-scale=1" />
        <Meta />
        <Links />
      </head>
      <body>
        <ClientOnly> <!-- the children of this component will not be rendered server-side -->
          <Outlet />
        </ClientOnly>
        <ScrollRestoration />
        <Scripts />
        <LiveReload />
      </body>
    </html>
  );
}
```

This will make FluentUI work instantly (at least it did for me), especially if you don't have custom themes, props-dependent component styles or custom styled components.

Additionally, make sure your Remix routes provide a `LoaderFunction` (and probably `ActionFunction`) returning a JSON response. You could still use Remix's helpers like `useLoaderData`, `useActionData` and `useFetcher` to let Remix use the Fetch API to exchange the JSON bits between client and server.

# Final notice

Remix has a different behavior for constant variables. In dev mode, the internal `require` cache will be reset after every request ([reference](https://github.com/remix-run/remix/discussions/2198#discussioncomment-2290735)). In production, this will not apply. Simply spoken, you can assume a behavior comparable to call `resetIds` before every request in development whereas in production `resetIds` is not called at all.

HTH,
Sebastian
